src/execution/index/art/art.cpp

```diff
@@ -988,67 +978,160 @@ void ART::CheckConstraintsForChunk(DataChunk &input, ConflictManager &conflict_m
+	ARTNode::Vacuum(*this, tree, vacuum_flags);
+
+	// finalize the vacuum operation
+	FinalizeVacuum(allocators, vacuum_flags);
+	UpdateMemoryUsage();
+}
+
 //===--------------------------------------------------------------------===//
 // Merging
 //===--------------------------------------------------------------------===//
 
+vector<idx_t> ART::InitializeMerge(vector<FixedSizeAllocator *> &allocators) {
+
+	vector<idx_t> buffer_counts;
+	buffer_counts.reserve(allocators.size());
+	for (auto &allocator : allocators) {
+		buffer_counts.emplace_back(allocator->buffers.size());
+	}
+	return buffer_counts;
+}
+
 bool ART::MergeIndexes(IndexLock &state, Index *other_index) {
 
 	auto other_art = (ART *)other_index;
+	auto allocators = GetAllocators();
+	auto other_allocators = other_art->GetAllocators();
 
-	if (!this->tree) {
-		IncreaseMemorySize(other_art->memory_size);
-		tree = other_art->tree;
-		other_art->tree = nullptr;
-		return true;
+	if (tree) {
+		//  fully deserialize other_index, and traverse it to increment its buffer IDs
+		auto buffer_counts = InitializeMerge(allocators);
+		other_art->tree.InitializeMerge(*other_art, buffer_counts);
 	}
 
-	return Node::MergeARTs(this, other_art);
+	// merge the node storage
+	for (idx_t i = 0; i < allocators.size(); i++) {
+		allocators[i]->Merge(*other_allocators[i]);
+	}
+
+	// merge the ARTs
+	if (!tree.Merge(*this, other_art->tree)) {
+		return false;
+	}
+	return true;
 }
 
 //===--------------------------------------------------------------------===//
 // Utility
 //===--------------------------------------------------------------------===//
 
 string ART::ToString() {
+	string str = "[empty]";
 	if (tree) {
-		return tree->ToString(*this);
+		str = tree.ToString(*this);
 	}
-	return "[empty]";
+	UpdateMemoryUsage();
+	return str;
 }
 
-void ART::Verify() {
-#ifdef DEBUG
-	idx_t current_mem_size = 0;
-	if (tree) {
-		current_mem_size = tree->MemorySize(*this, true);
-	}
-	if (memory_size != current_mem_size) {
-		throw InternalException("Memory_size value (%d) does not match actual memory size (%d).", memory_size,
-		                        current_mem_size);
-	}
-#endif
+vector<FixedSizeAllocator *> ART::GetAllocators() const {
+
+	vector<FixedSizeAllocator *> allocators;
+
+	allocators.push_back(prefix_segments.get());
+	allocators.push_back(leaf_segments.get());
+	allocators.push_back(leaves.get());
+	allocators.push_back(n4_nodes.get());
+	allocators.push_back(n16_nodes.get());
+	allocators.push_back(n48_nodes.get());
+	allocators.push_back(n256_nodes.get());
+
+	return allocators;
 }
 
-void ART::IncreaseAndVerifyMemorySize(idx_t old_memory_size) {
-	// since we lazily deserialize ART nodes, it is possible that its in-memory size
-	// increased during lookups
-	Verify();
-	D_ASSERT(memory_size >= old_memory_size);
-	if (track_memory) {
-		buffer_manager.IncreaseUsedMemory(memory_size - old_memory_size);
+void ART::UpdateMemoryUsage() {
+
+	idx_t current_size = 0;
```

Comment:
If the `FixedSizeAllocator` is updated to allocate through the buffer allocator the buffer manager should automatically keep track of the used memory, and we should be able to remove this method